### PR TITLE
[BEAM-1117] Support user timers for ParDo in the direct runner

### PR DESCRIPTION
**Pull request by kennknowles**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

R: @tgroh

### CODE REVIEW

**Comment by kennknowles** [comment]
OK, this is working superficially anyhow. It surely needs more tests. I will concoct as many as I can to be independent of execution strategy, so they can be cross-runner, but will resort to `TestStream` (direct runner only) for others, which will get coverage the shared utilities.

**Comment by kennknowles** [comment]
Confirmed that the fix to the windowing strategy that enables timers breaks state :-/

**Comment by kennknowles** [comment]
PTAL. Base PRs are merged and this is now focused on just exposing enough plumbing to deliver the timers.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/DoFnLifecycleManagerRemovingTransformEvaluator.java [hunk 1]
```diff
@@ -53,6 +55,15 @@ public void processElement(WindowedValue<InputT> element) throws Exception {
     }
   }
 
+  public void onTimer(TimerData timer, BoundedWindow window) throws Exception {
```
Comment (tgroh) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/DoFnLifecycleManagerRemovingTransformEvaluator.java]: There probably should be a JIRA to make this consistent with `SplittableDoFn` and `GroupAlsoByWindow`

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 2]
```diff
@@ -1672,6 +1675,75 @@ public void onTimer(OnTimerContext context) {
...
+            context.output(3);
+          }
+
+          @OnTimer(timerId)
+          public void onTimer(OnTimerContext context) {
+            context.output(42);
+          }
+        };
+
+    TestStream<KV<String, Integer>> stream = TestStream.create(KvCoder
```
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Formatting is funky.

Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Fixed.

**Pull request merged**